source/h2_decoder.c

```diff
@@ -456,6 +466,19 @@ static int s_state_fn_prefix(struct aws_h2_decoder *decoder, struct aws_byte_cur
     decoder->frame_in_progress.flags.priority =
         flags & AWS_H2_FRAME_F_PRIORITY || decoder->frame_in_progress.type == AWS_H2_FRAME_T_PRIORITY;
 
+    /* Connection preface requires that SETTINGS be sent first (RFC-7540 3.5).
+     * This should be the first error we check for, so that a connection sending
+     * total garbage data is likely to trigger this PROTOCOL_ERROR */
+    if (!decoder->connection_preface_complete) {
+        if (frame->type == AWS_H2_FRAME_T_SETTINGS && !frame->flags.ack) {
+            DECODER_LOG(TRACE, decoder, "Connection preface satisfied.");
+            decoder->connection_preface_complete = true;
+        } else {
+            DECODER_LOG(ERROR, decoder, "First frame must be SETTINGS");
+            return aws_raise_error(AWS_ERROR_HTTP_PROTOCOL_ERROR);
+        }
+    }
+
     /* Validate the frame's stream ID.
      * Frame types generally either require a stream-id, or require that it be zero. */
     frame->stream_id &= s_31_bit_mask; /* Mask off reserved bit */
```

Comment:
Should we fail if the high bit is set?  More generally, when the spec says "MUST NOT DO X" but doesn't explicitly define X as resulting in a (protocol/connection) error, should it be ignored or connection-terminal?

Reply:
> R: A reserved 1-bit field. The semantics of this bit are undefined, and the bit MUST remain unset (0x0) when sending and **MUST be ignored when receiving**

I'll add a comment to clarify